mathics/builtin/arithfns/basic.py

```diff
@@ -555,6 +555,7 @@ def apply_check(self, x, y, evaluation):
                 y_err = y
             else:
                 y_err = apply_N(y, evaluation)
+                y_err = y_err if y_err.is_atom() else y_err.evaluate(evaluation)
```

Comment:
I don't understand why we need to re-evaluate here and in other places.

Reply:
I have explained that in the docstring of `apply_N`. In the previous implementation,
```
Expression(SymbolN, expr, prec).evaluate(evaluation)
```
first, ``expr`` is evaluated. Then, ``apply_with_prec()`` is called and evaluated, and then, the result is returned to  the `evaluate` method. If there were a change in the expression, then the returned value is reevaluated.
In the new implementation, this would be equivalent to
````
apply_N(expr.evaluate(evaluation), prec.evaluate(evaluation)).evaluate(evaluation)
```
There are then several  improvements using `apply_N` instead of the previous construction:

* Inside a Python routine, usually `expr` is written in its final form (if you want to pass  `expr=F[4]`, you do not construct it as `expr=F[1+3]`).
* The return value of `apply_N` are usually `Real` or `Complex` numbers. Then, if you call `evaluation(evaluate)` you would be wasting a function call. If the result of `apply_N` is not a number, a comparison shouldn't introduce a substantial overhead over the work of a full evaluation (implying pattern matching of rules, and several function calls). On the other hand, if you get a number,  a comparison is always cheaper than a function call.